File: 107. Binary Tree Level Order Traversal II—Breadth-first Search/107. Binary Tree Level Order Traversal II—Breadth-first Search.java
class Solution {
    public List<List<Integer>> traverse(TreeNode root, int level, List<List<Integer>> res) {
		if (res.size() <= level && (root.left != null || root.right != null))
			res.add(0, new LinkedList<>());
		if (root.left != null) {
			res.get(res.size() - level - 1).add(root.left.val);
			traverse(root.left, level + 1, res);
		}
		if (root.right != null) {
			res.get(res.size() - level - 1).add(root.right.val);
			traverse(root.right, level + 1, res);
		}
		return res;
	}
	
	public List<List<Integer>> levelOrderBottom(TreeNode root) {
		List<List<Integer>> res = new LinkedList<>();
		if (root == null) return res;
		List<Integer> item = new LinkedList<>();
		item.add(root.val);
		res.add(item);
		res = traverse(root, 1, res);
		return res;
	}
}

File: 107. Binary Tree Level Order Traversal II—Breadth-first Search/107. Binary Tree Level Order Traversal II—Breadth-first Search.cpp
class Solution {
public:
    vector<vector<int>> levelOrderBottom(TreeNode* root) {
        vector<vector<int>>v;
        stack<vector<int>>s;
        queue<TreeNode*>q;
        q.push(root);  //根节点入队
        if(root == NULL)  
            return v ;
        while(!q.empty()){  //队列不空
            vector<int>vv;
            queue<TreeNode*> next ;  //  建立第二个队列 用来存放下一层的结点
            
            while(!q.empty()){  //遍历每层的结点  这层循环是核心 其他都是为了满足OJ输出
                
                TreeNode* tre = q.front() ;
                vv.push_back(tre->val);  //访问该结点,为了满足输出要求，所以有点复杂，
                q.pop();  //对头元素出队
                if(tre->left!=NULL){  //它有左子树
                    next.push(tre->left);
                }
                if(tre->right!=NULL){  //它有右子树
                    next.push(tre->right);
                }
                
            }
            s.push(vv);  //将每层结点入栈
            //v.push_back(vv);
            q=next;  // // 遍历完后进入下一层 
        }
        while(!s.empty()){  //将每层结点倒序输出
            v.push_back(s.top());
            s.pop();
        }
        return v;
    }
};